sdk/core/azure-core/src/http/response.cpp

```diff
@@ -40,18 +40,61 @@ inline bool IsStringEqualsIgnoreCase(std::string const& a, std::string const& b)
 }
 } // namespace
 
+void Response::AddHeader(std::string const& header)
+{
+  // get name and value from header
+  auto start = header.begin();
+  auto end = std::find(start, header.end(), ':');
+
+  if (end == header.end())
+  {
+    return; // not a valid header or end of headers symbol reached
+  }
+
+  auto headerName = std::string(start, end);
+  start = end + 1; // start value
+  while (start < header.end() && (*start == ' ' || *start == '\t'))
+  {
+    ++start;
+  }
+
+  end = std::find(start, header.end(), '\r');
+  auto headerValue = std::string(start, end); // remove \r
+
+  AddHeader(headerName, headerValue);
+}
+
 void Response::AddHeader(std::string const& name, std::string const& value)
 {
   if (IsStringEqualsIgnoreCase("Content-Length", name))
   {
-    // whenever this header is found, we reserve the value of it to be pre-allocated to write
-    // response
-    m_bodyBuffer.reserve(std::stol(value));
+    if (this->m_bodyStream != nullptr)
+    {
+      // Create an empty stream just to hold the size. Then any transport can replace for an
+      // specific stream, like libcurl, it will get the size and creates a CurlBodyStream
+      SetBodyStream(new MemoryBodyStream(nullptr, std::stol(value)));
+    }
+    else
+    {
+      // whenever this header is found, we reserve the value of it to be pre-allocated to write
+      // response
+      m_bodyBuffer.reserve(std::stol(value));
+    }
   }
   this->m_headers.insert(std::pair<std::string, std::string>(name, value));
 }
 
 void Response::AppendBody(uint8_t* ptr, uint64_t size)
 {
   m_bodyBuffer.insert(m_bodyBuffer.end(), ptr, ptr + size);
-}
\ No newline at end of file
+}
+
+void Response::SetBodyStream(BodyStream* stream)
+{
+  // Before setting body Stream, avoid leaking
+  if (this->m_bodyStream != nullptr)
+  {
+    delete this->m_bodyStream;
```

Comment:
I think it's also possible that this old `m_bodyStream` is created by user and passed in from `Request`. If this is the case, I don't think we should delete it. Because we have no idea if this stream is allocated on heap or stack, and we don't know if it's owned by a shared pointer.

Reply:
Not really. User can't create one stream for a Response. 
Users can create one Stream for a Request. Then we use that stream to read the request body and upload it to network.
But for the response, it is only the SDK the only one who is ever going to create the stream for a Response as the way for a user to read the response body lazily.

Reply:
Then how about a user creates one(or two) BodyStream and call this function twice?

Reply:
Deleting the input stream doesn't sound right here. There should be no assumption made on the input stream's status, the only thing we should do is to write to this stream.

Reply:
Yeah, you are right.
We should not delete the stream. It is responsibility of whoever that creates the HTTP Request to take care of the stream.

Only for the stream that is in the HTTP Response, the responsibility is from the HTTP Response, to avoid leaking.